src/service_control_client_impl_test.h

```diff
@@ -0,0 +1,893 @@
+      CheckResponse* transport_response) {
+    EXPECT_CALL(mock_check_transport_, Check(_, _, _))
+        .WillOnce(Invoke(&mock_check_transport_,
+                         &MockCheckTransport::CheckWithInplaceCallback));
+
+    // Set the check status and response to be used in the on_check_done
+    mock_check_transport_.done_status_ = transport_status;
+    mock_check_transport_.check_response_ = transport_response;
+
+    CheckResponse check_response;
+    Status done_status = client_->Check(request, &check_response);
+
+    EXPECT_EQ(done_status, transport_status);
+    EXPECT_TRUE(MessageDifferencer::Equals(mock_check_transport_.check_request_,
+                                           request));
+    if (transport_status.ok()) {
+      EXPECT_TRUE(
+          MessageDifferencer::Equals(*transport_response, check_response));
+    }
+
+    // Verifies call expectations and clear it before other test.
+    EXPECT_TRUE(Mock::VerifyAndClearExpectations(&mock_check_transport_));
+  }
+
+  // Tests non cached check request. Mocked transport::Check() is using a
+  // separate thread to call on_done callback.
+  // 1) Call a Client::Check(),  the request is not in the cache.
+  // 2) Transport::Check() is called. Mocked transport::Check() creates a
+  //    separate thread to call on_done callback.
+  // 3) Client::Check() returns, but Client::on_check_done() will be called
+  //    from the other thread.
+  void InternalTestNonCachedCheckUsingThread(
+      const CheckRequest& request, Status transport_status,
+      CheckResponse* transport_response) {
+    EXPECT_CALL(mock_check_transport_, Check(_, _, _))
+        .WillOnce(Invoke(&mock_check_transport_,
+                         &MockCheckTransport::CheckUsingThread));
+
+    // Set the check status and response to be used in the on_check_done
+    mock_check_transport_.done_status_ = transport_status;
+    mock_check_transport_.check_response_ = transport_response;
+
+    StatusPromise status_promise;
+    StatusFuture status_future = status_promise.get_future();
+
+    CheckResponse check_response;
+    client_->Check(request, &check_response, [&status_promise](Status status) {
+      StatusPromise moved_promise(std::move(status_promise));
+      moved_promise.set_value(status);
+    });
+
+    // Since it is not cached, transport should be called.
+    EXPECT_TRUE(MessageDifferencer::Equals(mock_check_transport_.check_request_,
+                                           request));
+
+    // on_check_done is called with right status.
+    status_future.wait();
+    EXPECT_EQ(transport_status, status_future.get());
+    if (transport_status.ok()) {
+      EXPECT_TRUE(
+          MessageDifferencer::Equals(*transport_response, check_response));
+    }
+
+    // Verifies call expections and clear it before other test.
+    EXPECT_TRUE(Mock::VerifyAndClearExpectations(&mock_check_transport_));
+  }
+
+  void InternalTestNonCachedBlockingCheckUsingThread(
```

Comment:
it seems there are not much code in this file can be shared between Check/Report.  
Maybe it is better to split this header file into its own _chec_test.cc and _report_test.cc